src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -0,0 +1,264 @@
+                }
+            }
+            catch (Exception ex)
+            {
+                if (Log.IsEnabled(LogEventLevel.Error))
+                {
+                    Log.Error(ex, "Error extracting propagated HTTP headers.");
+                }
+            }
+
+            return null;
+        }
+
+        private void OnMvcAfterOnActionExecuted(object arg)
+        {
+            if (AfterOnActionActionExecutedContextFetcher.Fetch(arg) is ActionExecutedContext actionExecutedContext &&
+                actionExecutedContext.Exception != null &&
+                !actionExecutedContext.ExceptionHandled)
+            {
+                _tracer.ScopeManager.Active?.Span.SetException(actionExecutedContext.Exception);
+            }
+        }
+
+        private void OnHostingHttpRequestInStop(object arg)
+        {
+            IScope scope = _tracer.ScopeManager.Active;
+
+            if (scope != null)
+            {
+                var httpContext = (HttpContext)HttpRequestInStopHttpContextFetcher.Fetch(arg);
+                scope.Span.SetTag(Tags.HttpStatusCode, httpContext.Response.StatusCode.ToString());
+
+                if (httpContext.Response.StatusCode / 100 == 5)
+                {
+                    // 5xx codes are server-side errors
+                    scope.Span.Error = true;
+                }
+
+                scope.Dispose();
+            }
+        }
+
+        private void OnMvcBeforeAction(object arg)
+        {
+            var httpContext = (HttpContext)BeforeActionHttpContextFetcher.Fetch(arg);
+
+            if (ShouldIgnore(httpContext))
+            {
+                if (Log.IsEnabled(LogEventLevel.Debug))
+                {
+                    Log.Debug("Ignoring request");
+                }
+            }
+            else
+            {
+                // NOTE: This event is the start of the action pipeline. The action has been selected, the route
+                //       has been selected but no filters have run and model binding hasn't occured.
+                var actionDescriptor = (ActionDescriptor)BeforeActionActionDescriptorFetcher.Fetch(arg);
+                var controllerActionDescriptor = actionDescriptor as ControllerActionDescriptor;
+                HttpRequest request = httpContext.Request;
+
+                string httpMethod = request.Method?.ToUpperInvariant() ?? "UNKNOWN";
+                string controllerName = controllerActionDescriptor?.ControllerName;
+                string actionName = controllerActionDescriptor?.ActionName;
+                string routeTemplate = actionDescriptor.AttributeRouteInfo?.Template ?? $"{controllerName}/{actionName}";
+                string resourceName = $"{httpMethod} {routeTemplate}";
+
+                Span span = _tracer.ScopeManager.Active?.Span;
+
+                // override the parent's resource name with the MVC route template
+                if (span != null)
```

Comment:
Maybe you could move this span check earlier so we don't do as much unnecessary work

Reply:
Definitely, thanks. Changed in 900852154eb0d35803c4e32946e65a769bb8d421.